File: kotlin/test/PricesTest.kt
import dojo.Prices
import io.ktor.server.engine.*
import io.restassured.RestAssured
import io.restassured.path.json.JsonPath
import io.restassured.response.ResponseBodyExtractionOptions
import io.restassured.specification.RequestSpecification
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.sql.Connection
import kotlin.test.assertEquals

internal class PricesTest {
    private lateinit var connection: Connection
    private lateinit var app: ApplicationEngine

    @BeforeEach
    fun createPrices() {
        Prices.createApp().let {
            connection = it.first
            app = it.second
        }
        app.start()
    }

    @ParameterizedTest
    @CsvSource("1jour, 35",
            "night, 19")
    fun updateDefaultPrice(type: String, cost: Int) {
        createPrice(type, cost)
    }

    private fun createPrice(type: String, cost: Int) {
        given().
            `when`().
            params("type", type, "cost", cost).
            put("/prices").
        then().
            assertThat().
                contentType("application/json").
            assertThat().
                statusCode(200) // TODO should be 204
    }

    @AfterEach
    fun stopApplication() {
        app.stop(200, 200)
        connection.close()
    }

    @Test
    fun defaultCost() {
        val json = obtainPrice("type", "1jour")
        val cost = json.get<Int>("cost")
        assertEquals(35, cost)
    }

    @ParameterizedTest
    @CsvSource("5, 0",
            "6, 25",
            "14, 25",
            "15, 35",
            "25, 35",
            "64, 35",
            "65, 27") //
    fun worksForAllAges(age: String, expectedCost: Int) {
        val json = obtainPrice("type", "1jour", "age", age)
        val cost = json.get<Int>("cost")
        assertEquals(expectedCost, cost)
    }

    @Test
    fun realNightCost() {
        val json = obtainPrice("type", "night")
        val cost = json.get<Int>("cost")
        assertEquals(0, cost)
    }

    @Test
    @Disabled
    fun defaultNightCost() {
        val json = obtainPrice("type", "night")
        val cost = json.get<Int>("cost")
        assertEquals(19, cost)
    }

    @ParameterizedTest
    @CsvSource("5, 0",
            "6, 19",
            "25, 19",
            "64, 19",
            "65, 8") //
    fun worksForNightPasses(age: Int, expectedCost: Int) {
        val json = obtainPrice("type", "night", "age", age)
        val cost = json.get<Int>("cost")
        assertEquals(expectedCost, cost)
    }

    @ParameterizedTest
    @CsvSource("15, '2019-02-22', 35", 
            "15, '2019-02-25', 35",
            "15, '2019-03-11', 23",
            "65, '2019-03-11', 18") //
    fun worksForMondayDeals(age: Int, date: String, expectedCost: Int) {
        val json = obtainPrice("type", "1jour", "age", age, "date", date)
        val cost = json.get<Int>("cost")
        assertEquals(expectedCost, cost)
    }

    // TODO 2-4, and 5, 6 day pass

    private fun given(): RequestSpecification {
        return RestAssured.given().
            accept("application/json").
            port(4567)
    }

    private fun obtainPrice(paramName: String, paramValue: Any, vararg otherParamPairs: Any): JsonPath {
        return given().
            `when`().
                params(paramName, paramValue,  *otherParamPairs).
                get("/prices").
            then().
                assertThat().
                    contentType("application/json").
                assertThat().
                    statusCode(200).
                extract().jsonPath()
    }

    protected fun RequestSpecification.When(): RequestSpecification {
        return this.`when`()
    }

    // allows response.to<Widget>() -> Widget instance
    protected inline fun <reified T> ResponseBodyExtractionOptions.to(): T {
        return this.`as`(T::class.java)
    }
}

File: csharp-simple/LiftPassPricingTests/PricesTest.cs
using System;
using Xunit;
using Nancy;
using Nancy.Testing;
using LiftPassPricing;
using System.Collections.Generic;

namespace LiftPassPricingTests
{
    /// <seealso>"http://www.marcusoft.net/2013/01/NancyTesting1.html"</seealso>
    public class PricesTest : IDisposable
    {
        private readonly Prices prices;
        private readonly Browser browser;

        public PricesTest()
        {
            this.prices = new Prices();
            this.browser = new Browser(with => with.Module(prices));

            CreatePrice("1jour", 35);
            CreatePrice("night", 19);
        }

        private void CreatePrice(string type, int cost)
        {
            var result = browser.Put("/prices", with =>
            {
                with.Query("type", type);
                with.Query("cost", cost.ToString());
                with.HttpRequest();
            });

            Assert.Equal("application/json", result.Result.ContentType);
            Assert.Equal(HttpStatusCode.OK, result.Result.StatusCode); // TODO should be 204
        }

        public void Dispose()
        {
            prices.connection.Close();
        }

        [Fact]
        public void DefaultCost()
        {
            Response json = ObtainPrice("type", "1jour");
            Assert.Equal(35, json.Cost);
        }

        [Theory]
        [InlineData(5, 0)]
        [InlineData(6, 25)]
        [InlineData(14, 25)]
        [InlineData(15, 35)]
        [InlineData(25, 35)]
        [InlineData(64, 35)]
        [InlineData(65, 27)]
        public void WorksForAllAges(int age, int expectedCost)
        {
            Response json = ObtainPrice("type", "1jour", "age", age.ToString());
            Assert.Equal(expectedCost, json.Cost);
        }

        [Theory]
        [InlineData(5, 0)]
        [InlineData(6, 19)]
        [InlineData(25, 19)]
        [InlineData(64, 19)]
        [InlineData(65, 8)]
        public void WorksForNightPasses(int age, int expectedCost)
        {
            Response json = ObtainPrice("type", "night", "age", age.ToString());
            Assert.Equal(expectedCost, json.Cost);
        }

        [Theory]
        [InlineData(15, "2019-02-22", 35)]
        [InlineData(15, "2019-02-25", 35)]
        [InlineData(15, "2019-03-11", 23)]
        [InlineData(65, "2019-03-11", 18)]
        public void WorksForMondayDeals(int age, string date, int expectedCost)
        {
            Response json = ObtainPrice("type", "1jour", "age", age.ToString(), "date", date);
            Assert.Equal(expectedCost, json.Cost);
        }

        // TODO 2-4, and 5, 6 day pass

        private Response ObtainPrice(params string[] keyValuePairs)
        {
            var result = browser.Get("/prices", with =>
            {
                for (int i = 0; i < keyValuePairs.Length; i += 2)
                {
                    with.Query(keyValuePairs[i], keyValuePairs[i + 1]);
                }
                with.HttpRequest();
            });

            Assert.Equal("application/json", result.Result.ContentType);
            Assert.Equal(HttpStatusCode.OK, result.Result.StatusCode);

            return result.Result.Body.DeserializeJson<Response>();
        }
    }

    class Response
    {
        public int Cost { get; set; }
    }
}
